src/dataflow-types/src/types.rs

```diff
@@ -1307,42 +1313,115 @@ pub mod sources {
+                    RegionProviderChain::first_try(Region::new(region.clone()))
+                        .or_default_provider()
+                } else {
+                    let mut wc =
+                        aws_config::default_provider::region::DefaultRegionChain::builder();
+                    if let AwsCredentials::Profile { profile_name } = &self.credentials {
+                        wc = wc.profile_name(profile_name);
+                    }
+                    // TODO: this doesn't work for profiles which do not provide their own region
+                    // but instead specify a source_profile that does have a region
+                    RegionProviderChain::first_try(wc.build()).or_default_provider()
+                }
+            };
+
+            loader = loader.region(region());
+
+            let cred_provider = match &self.credentials {
+                AwsCredentials::Default => SharedCredentialsProvider::new(
+                    DefaultCredentialsChain::builder()
+                        .region(region())
+                        .build()
+                        .await,
+                ),
+                AwsCredentials::Profile { profile_name } => SharedCredentialsProvider::new(
+                    DefaultCredentialsChain::builder()
+                        .profile_name(profile_name)
+                        .region(region())
+                        .build()
+                        .await,
+                ),
+                AwsCredentials::Static {
+                    access_key_id,
+                    secret_access_key,
+                    session_token,
+                } => SharedCredentialsProvider::new(aws_types::Credentials::from_keys(
+                    access_key_id,
+                    secret_access_key,
+                    session_token.clone(),
+                )),
+            };
+
+            let role = if let Some(role_arn) = self.role_arn.as_ref() {
+                let mut role = AssumeRoleProvider::builder(role_arn);
+
+                // this affects which region to perform STS on, not where anything else happens
+                if let Some(region) = region().region().await {
+                    role = role.region(region);
+                }
+                if let Some(external_id) = &self.external_id {
+                    role = role.external_id(external_id);
+                }
+                role = role.session_name("materialized");
+
+                Some(role)
+            } else {
+                None
+            };
+
+            loader = if let Some(role) = role {
+                loader.credentials_provider(role.build(cred_provider))
+            } else {
+                loader.credentials_provider(cred_provider)
+            };
+
             let mut config = mz_aws_util::config::AwsConfig::from_loader(loader).await;
             if let Some(endpoint) = &self.endpoint {
                 config.set_endpoint(Endpoint::immutable(endpoint.0.clone()));
```

Comment:
So I think this whole function gets a good bit clearer if you structure it like so:

```rust
        /// Loads the AWS SDK configuration object from the environment, then
        /// applies the overrides from this object.
        pub async fn load(&self) -> mz_aws_util::config::AwsConfig {
            use aws_config::default_provider::credentials::DefaultCredentialsChain;
            use aws_config::default_provider::region::DefaultRegionChain;
            use aws_config::sts::AssumeRoleProvider;
            use aws_smithy_http::endpoint::Endpoint;
            use aws_types::credentials::SharedCredentialsProvider;
            use aws_types::region::Region;

            let region = match &self.region {
                Some(region) => Some(Region::new(region.clone())),
                _ => {
                    let mut rc = DefaultRegionChain::builder();
                    if let AwsCredentials::Profile { profile_name } = &self.credentials {
                        rc = rc.profile_name(profile_name);
                    }
                    // TODO: this doesn't work for profiles which do not provide their own region
                    // but instead specify a source_profile that does have a region
                    rc.build().region().await
                }
            };

            let mut cred_provider = match &self.credentials {
                AwsCredentials::Default => SharedCredentialsProvider::new(
                    DefaultCredentialsChain::builder()
                        .region(region.clone())
                        .build()
                        .await,
                ),
                AwsCredentials::Profile { profile_name } => SharedCredentialsProvider::new(
                    DefaultCredentialsChain::builder()
                        .profile_name(profile_name)
                        .region(region.clone())
                        .build()
                        .await,
                ),
                AwsCredentials::Static {
                    access_key_id,
                    secret_access_key,
                    session_token,
                } => SharedCredentialsProvider::new(aws_types::Credentials::from_keys(
                    access_key_id,
                    secret_access_key,
                    session_token.clone(),
                )),
            };

            if let Some(role_arn) = &self.role_arn {
                let mut role = AssumeRoleProvider::builder(role_arn).session_name("materialized");
                // This affects which region to perform STS on, not where
                // anything else happens.
                if let Some(region) = &region {
                    role = role.region(region.clone());
                }
                if let Some(external_id) = &self.external_id {
                    role = role.external_id(external_id);
                }
                cred_provider = SharedCredentialsProvider::new(role.build(cred_provider));
            }

            let loader = aws_config::from_env().region(region).credentials_provider(cred_provider);
            let mut config = mz_aws_util::config::AwsConfig::from_loader(loader).await;
            if let Some(endpoint) = &self.endpoint {
                config.set_endpoint(Endpoint::immutable(endpoint.0.clone()));
            }
            config
        }
```

Btw, I put the `use` block inside the function because most of the types (`Region`, `Endpoint`) aren't obviously AWS types, and it didn't seem right to import them into the whole massive module's scope.